interpreter/text/arrange.ml

```diff
@@ -405,23 +416,23 @@ let module_ = module_with_var_opt None
 
 (* Scripts *)
 
+(* Converts a value to string depending on mode. *)
 let literal mode lit =
+  let choose_mode bin not_bin = if mode = `Binary then bin else not_bin in
   match lit.it with
-  | Values.I32 i ->
-    let f = if mode = `Binary then I32.to_hex_string else I32.to_string_s in
-    Node ("i32.const " ^ f i, [])
-  | Values.I64 i ->
-    let f = if mode = `Binary then I64.to_hex_string else I64.to_string_s in
-    Node ("i64.const " ^ f i, [])
-  | Values.F32 z ->
-    let f = if mode = `Binary then F32.to_hex_string else F32.to_string in
-    Node ("f32.const " ^ f z, [])
-  | Values.F64 z ->
-    let f = if mode = `Binary then F64.to_hex_string else F64.to_string in
-    Node ("f64.const " ^ f z, [])
-  | Values.V128 v -> 
-    let f = if mode = `Binary then (failwith "unimplemented v128 binary mode") else V128.to_string in
-    Node ("v128.const " ^ f v, [])
+  | Values.I32 i -> (choose_mode I32.to_hex_string I32.to_string_s) i
+  | Values.I64 i -> (choose_mode I64.to_hex_string I64.to_string_s) i
+  | Values.F32 z -> (choose_mode F32.to_hex_string F32.to_string) z
+  | Values.F64 z -> (choose_mode F64.to_hex_string F64.to_string) z
+  | Values.V128 v -> (choose_mode V128.to_hex_string V128.to_string) v
+
+(* Converts a literal into a constant instruction. *)
+let constant mode lit =
+  let type_name = string_of_value_type (type_of lit.it) in
+  let lit_string = literal mode lit in
+  match lit.it with
+  | Values.V128 v -> Node ("v128.const i32x4 " ^ lit_string, [])
```

Comment:
This logic to display " i32x4 " comes up a few times, is it possible to use the SIMD shape representation here? Perhaps as `shape option` to handle non-SIMD types?

Reply:
Not sure if we want the "shape" logic inside of all the numeric const nodes, we might have to come up with our own SimdConst node, then it can have a Shape. I can probably try that out as a follow up in order to have more consistent translation between text and binary, and keep the shape information.
Does Wabt do this? (shape option in Const AST nodes)

Reply:
I was thinking that the shape would be a parameter, not necessarily stored in the node. But yeah, doesn't have to be this PR

Reply:
Ah okay, I don't think it's possible here, because constant here is called by instructions that don't have the knowledge of the shape.
If you look below, `let result mode res =` on line 498, we branch on `SimdResult` or `NumResult`, we only have the shape in the first case, in the second case we end up with a V128 value, and have no information on the shape.

Reply:
Right, I guess I mean for clarity so there doesn't have to be an explicit match -- so the shape could still be a constant `I32X4`, but the conversion to text is moved to a function.

Reply:
I moved this check into `constop` above, so this "i32x4" lives in a single place now.